src/librustc/ty/subst.rs

```diff
@@ -239,27 +239,15 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
                            defs: &ty::Generics,
                            mk_region: &mut FR,
                            mk_type: &mut FT)
-    where FR: FnMut(&ty::RegionParameterDef, &[Kind<'tcx>]) -> ty::Region<'tcx>,
-          FT: FnMut(&ty::TypeParameterDef, &[Kind<'tcx>]) -> Ty<'tcx> {
-        // Handle Self first, before all regions.
-        let mut types = defs.types.iter();
-        if defs.parent.is_none() && defs.has_self {
-            let def = types.next().unwrap();
-            let ty = mk_type(def, substs);
-            assert_eq!(def.index as usize, substs.len());
-            substs.push(ty.into());
-        }
-
-        for def in &defs.regions {
-            let region = mk_region(def, substs);
-            assert_eq!(def.index as usize, substs.len());
-            substs.push(Kind::from(region));
-        }
-
-        for def in types {
-            let ty = mk_type(def, substs);
-            assert_eq!(def.index as usize, substs.len());
-            substs.push(Kind::from(ty));
+    where FR: FnMut(&ty::GenericParamDef, &[Kind<'tcx>]) -> ty::Region<'tcx>,
+          FT: FnMut(&ty::GenericParamDef, &[Kind<'tcx>]) -> Ty<'tcx> {
+        for param in &defs.params {
+            let kind = match param.kind {
+                ty::GenericParamDefKind::Lifetime => mk_region(param, substs).into(),
+                ty::GenericParamDefKind::Type(_) => mk_type(param, substs).into(),
```

Comment:
I almost want to ask you to combine `mk_region` and `mk_type` and make users pattern-match `.kind`, but let's not drag this even further.

Reply:
Yeah, I wanted to do this too. I'll do it in a follow-up once this PR's been merged.

Reply:
I've now made this change in a separate branch. If you'd like to have it included in this PR, I can merge it in. If not, I'll make a separate PR later.